sdk/ml/azure-ai-ml/azure/ai/ml/entities/_job/pipeline/_attr_dict.py

```diff
@@ -113,14 +113,14 @@ def __getattr__(self, key: K) -> V:
             self.__setattr__(key, result)
             return result
 
-    def __setattr__(self, key: K, value: V):
+    def __setattr__(self, key: K, value: V) -> None:
```

Comment:
When should we use None vs. NoReturn?

Reply:
[NoReturn](https://docs.python.org/3.10/library/typing.html#typing.NoReturn) is meant to signal that a function never returns normally. [Mypy's docs have a nice explainer](https://mypy.readthedocs.io/en/stable/more_types.html#the-noreturn-type)

> Mypy provides support for functions that never return. For example, a function that unconditionally raises an exception:
> ```python
> from typing import NoReturn
> 
> def stop() -> NoReturn:
>     raise Exception('no way')
> ```


It was useful in this PR since we have a couple functions that never return (e.g. raise formatted exceptions), and NoReturn lets me signal intent to pylint (e.g. https://github.com/Azure/azure-sdk-for-python/pull/31147/commits/f08abe1d545e14a047036fe802d8735f00986e73)